Initial Traktor MK2 vinyl/CD support

jclsn opened the pull request:
This PR adds initial Traktor MK2 support. Why initial? This timecode is very different from the Serato-style ones. The modulation technique is different and also the LFSR sequence itself has some other properties, which make processing of the timecode more demanding in terms of complexity, CPU load and memory requirements. I might be able to polish all of these things in the future, but for now I'm happy with basic support.

What works:

- [x] Timecode detection when using a proper phono preamp
- [x] Scope 
- [x] Side-A, Side-B and CD work
- [x] Fix a crash when reloading timecode definitions
- [x] LUT caching (Linux)
- [x]  LUT caching (Windows)
- [x]  LUT caching (macOS)

To do in later PRs:
- **Phono level timecode detection:** Due to the different waveform, without passing the signal through a phono preamp that applies RIAA De-Emphasis, the signal has different characteristics and another demodulation algorithm needs to be developed. 
- **LUT improvements:** The LUT needs quite a lot of memory atm (100MB for the CD) and the generation at startup takes about 15 seconds. The current the workaround is caching the LUT to disk and loading it into memory on startup. This way the startup time of Mixxx is not affected, unless you delete the ``.lut`` files, which would trigger a regeneration.

jclsn committed:
vinylcontrol: Add Traktor MK2s to timecode collection

Adds Traktor MK2 Side, A/B and CD

jclsn committed:
xwax: Add custom u128 struct for the sake of portability

Mark wishes to use a custom u128 type instead of C23 features like
arbitrary-sized integers using _BitInt() for the sake of portability and
performance.

jclsn committed:
xwax: Extend LUT with functions for the MK2 which can hold 110-bit LFSR states

Since the Traktor MK2 timecode is comprised of an LFSR, which generator
polynomial is of 110th order, a new type must be created to hold these
values.

This significantly increases the size of the LUT due to the type
being used in struct slot. Therefore the functions for the MK2 are
clearly separated to not make the current LUT gain in size by
holding unnecessarily padded zeros.

jclsn committed:
xwax: Consistently use bits_t and slot_no_t types in LUT

jclsn committed:
xwax: Add timecode definitions for Traktor MK2

Carefully worked out definitions for the Traktor MK2 timecodes.
Due to the high frequency of the carrier wave, the number of slots is
much higher.

jclsn committed:
xwax: Add needed filter structures

A simple exponential moving average (lowpass) is needed for smoothing
the signal before applying the derivative.

The derivative is needed to obtain a signal to feed to the pitch
detection.

The RMS value is needed to obtain an appropriate scaling factor for the
derivative.

jclsn committed:
xwax: Implement a ring buffer as delayline

The delayline is needed for computing the RMS value and for
demodulation.

Since the derivative and filter introduce a delay, this has to be taken
into account when getting timecode readings.

jclsn committed:
xwax: Push samples into the delayline

jclsn committed:
xwax: Compute the RMS value

Computes a smoothed RMS value to properly track the current signal
strength.

jclsn committed:
xwax: Implement pitch detection for Traktor MK2

The Traktor MK2 signal is offset-modulated and therefore not directly
suited for pitch detection.

A suitable signal is the derivative, which more or less evenly oscillates
around zero. Since the derivative's amplitude is only a fraction of the
amplitude of the original signal, it needs proper scaling.

Scaling is achieved via computing the RMS value of the running sums of
the signal and derivative. The resulting quotient of RMS_signal /
RMS_derivative yields the needed scaling factor.

jclsn committed:
xwax: Implement demodulation functions for Traktor MK2

ywwg commented:
As this PR comes together, we should determine what unit tests we can write to confirm that the code is working. That could involve a snippet (half second?) of each of the three timecode types and confirming that the code interprets them correctly. While I am happy to hand-test for now, true stability will come from being able to construct automatic tests

jclsn commented:
I wanted to get into testing anyway. Bought a book on it last year

JoergAtGithub commented:
Welcome at Mixxx!
As a first-time contributor we need you to sign the [Mixxx Contributor Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ) and comment here when you have done so. It gives us permission to distribute your contribution under the GPL v2 or later license and the Apple Mac App Store. It is also helpful for us to have contact information for contributors in case we may need it in the future.

jclsn commented:
@JoergAtGithub 

- [x] signed

JoergAtGithub commented:
Please have a look on the CI build fails

jclsn commented:
MVSC warns about C99 style. The whole xwax project is not consistently formatted or using modern C standards. I could the new ``__BitInt(110)`` feature for the 110-bit integer, but Mark wanted a customs struct for portability. I guess C and C++ bite each other here!

jclsn commented:
It's not merely style warnings. Seems like MSVC is failing when using C99 syntax. I tried enforcing compilation as C code but to no avail. Maybe enforcing C99 for xwax might help.

jclsn commented:
Yes, caching the LUT is possible. I already implemented something similar in xwax, but discarded it again. This would have to be done during the first run of Mixxx. Problem is that once you change something about the timecode definitions, the LUT would need to be regenerated.

ywwg commented:
A LUT version number is sufficient for deciding whether to generate new data.  worrying about data corruption on disk is out of scope and is not handled in any other component of mixxx, such as waveforms, overviews, the sqlite database, or the config. If such an issue arises, simply deleting the bad files would trigger regeneration.

jclsn commented:
The version number could also be part of the LUT object and compared on startup. If no match is found, the LUT is regenerated. Without automatic regeneration using a version number is pointless. I wouldn't want people to have to manually delete files. This would also be a source of many unnecessary issues and discussions in the forums.

ywwg commented:
I'll take another look soon

ywwg commented:
ready for review?

jclsn commented:
@ywwg Should be good! I just cleaned up the code and made sure that all commits compile.

jclsn commented:
I just realized there is still an issue on Windows. Putting back into draft.

jclsn commented:
Fixed the Windows issue.

jclsn commented:
> As this PR comes together, we should determine what unit tests we can write to confirm that the code is working. That could involve a snippet (half second?) of each of the three timecode types and confirming that the code interprets them correctly. While I am happy to hand-test for now, true stability will come from being able to construct automatic tests

But does any of this exist even for the other timecodes? What we could do is count the number of readings vs. correct readings and aim for BER > 0.95 or something. In any case this would involve uploading snippets of the timecodes into the repo, which not only would have copyright implications, but also would have to be a collective effort, because I'd assume no one owns copies of all timecodes.

ywwg commented:
I will review!

ywwg reviewed (changes requested):
My main note is architectural: given the amount of if/else switching depending on whether the timecoder is mk2 or not, I think it might be cleanest and easiest to maintain if mk2 is simply a separate set of files/objects that live apart from the existing timecoder.  More detailed notes inside the review.  Please tell me what you think.

jclsn commented on lib/xwax/timecoder.c:
```diff
@@ -336,6 +727,13 @@ void timecoder_init(struct timecoder *tc, struct timecode_def *def,
 void timecoder_clear(struct timecoder *tc)
 {
     assert(tc->mon == NULL);
+
+    if (tc->def->flags & TRAKTOR_MK2) {
+        delayline_init(&tc->primary.mk2.delayline);
+        delayline_init(&tc->secondary.mk2.delayline);
+        delayline_init(&tc->upper_subcode.readings);
+        delayline_init(&tc->lower_subcode.readings);
+    }
```
Well, the timecoder functions would still remain the same and I wouldn't want to initialize any of the MK2-related stuff for e.g. Serato. Mark was against having code duplication.

In this case the subcode structs live in the timecoder, because they are not channel-specific. The other quantities are channel-specific and since there are more of them, I put them into the struct mk2.

jclsn commented on lib/xwax/delayline.c:
```diff
@@ -0,0 +1,91 @@
+        delayline->size = DELAYLINE_SIZE;
+		delayline->current = delayline->size -1;
+        for (int i = 0; i < DELAYLINE_SIZE; i++)
+            delayline->array[i] = 0;
```
Changed

jclsn commented on lib/xwax/delayline.c:
```diff
@@ -0,0 +1,91 @@
+int delayline_avg(struct delayline *delayline)
+{
+    if (!delayline) {
+        printf("%s: Null pointer exception\n", __func__);
```
Using fprintf now for consistency.

jclsn commented on lib/xwax/timecoder.c:
```diff
@@ -170,6 +226,28 @@ static inline bits_t lfsr(bits_t code, bits_t taps)
+    }
+
+    return u128_and(xrs, U128_ONE);
+}
```
Put MK2-related functions into their own source file

jclsn commented on lib/xwax/timecoder.c:
```diff
@@ -495,8 +1006,43 @@ static void process_bitstream(struct timecoder *tc, signed int m)
 {
-    detect_zero_crossing(&tc->primary, primary, tc->zero_alpha, tc->threshold);
-    detect_zero_crossing(&tc->secondary, secondary, tc->zero_alpha, tc->threshold);
+    if (tc->def->flags & TRAKTOR_MK2) {
```
I grouped the functions that manipulate the carrier now into their own function now. I also put this function into the `timecoder_mk.c` for maintainability.

ywwg reviewed (commented):
I am going to try my best to review the C code

ywwg commented:
> Pull Request Overview

copilot is often stupid and wrong, but let's see if it has any good ideas :)

jclsn commented on lib/xwax/filters.c:
```diff
@@ -0,0 +1,96 @@
+    filter->squared_old = (1.0 - filter->alpha) * filter->squared_old + filter->alpha * squared;
+
+    // Take square root at the end
+    return (int) sqrt(filter->squared_old);
```
It needs to be long long for the computations, else it would overflow. Afterwards I convert back to the initial input type. I don't see a problem here.

jclsn commented on lib/xwax/filters.c:
```diff
@@ -0,0 +1,96 @@
+{
+    if (!filter) {
+        errno = EINVAL;
+        printf(__func__);
```
Yes, this should be `perror(__func__)`. Made a careless mistake here.

jclsn replied:
Changed to `perror(__func__`).

jclsn commented on lib/xwax/filters.c:
```diff
@@ -0,0 +1,96 @@
+{
+    if (!filter) {
+        errno = EINVAL;
+        printf(__func__);
```
Changed to `perror(__func__`).

jclsn commented on lib/xwax/filters.c:
```diff
@@ -0,0 +1,96 @@
+{
+    if (!filter) {
+        errno = EINVAL;
+        printf(__func__);
```
Changed to `perror(__func__`).

jclsn commented on lib/xwax/filters.c:
```diff
@@ -0,0 +1,96 @@
+{
+    if (!filter) {
+        errno = EINVAL;
+        printf(__func__);
```
Changed to `perror(__func__`).

jclsn commented on lib/xwax/timecoder_mk2.c:
```diff
@@ -0,0 +1,462 @@
+    tc->ref_level -= tc->ref_level / REF_PEAKS_AVG;
+    tc->ref_level += abs((int) (tc->secondary.mk2.rms_deriv * tc->gain_compensation)) / REF_PEAKS_AVG;
+
+    debug("upper.valid_counter: %d, lower.valid_counter %d, forwards: %b\n", */
```
Removed the leftover closing comment.

jclsn commented on src/vinylcontrol/vinylcontrolxwax.cpp:
```diff
@@ -192,6 +199,19 @@ void VinylControlXwax::freeLUTs() {
+    if (!lutPath.exists()) {
+        if (!lutPath.mkpath(".")) {
+            qWarning() << "Failed to create LUT directory at" << lutPath;
+            return nullptr;
```
Changed the code to correctly pass `nullptr` to `timecoder_find_definitions()` now if the config path is empty.

ywwg commented on lib/xwax/lut.h:
```diff
@@ -33,10 +42,22 @@ struct lut {
+void lut_push(struct lut *lut, bits_t timecode);
+unsigned int lut_lookup(struct lut *lut, bits_t timecode);
+
+int lut_init_mk2(struct lut_mk2 *lut, int nslots);
+void lut_clear_mk2(struct lut_mk2 *lut);
+
+void lut_push_mk2(struct lut_mk2 *lut, mk2bits_t *timecode);
+slot_no_t lut_lookup_mk2(struct lut_mk2 *lut, mk2bits_t *timecode);
```
So what I'm thinking is we separate all of the changes in this file into lut_traktor_mk2.h like you did with the timecoder file.  that will keep this file pristine and unedited from mark's work.  And the corresponding implementation can go into lut_traktor_mk2.c.

jclsn replied:
Put into lut_mk2.c

ywwg reviewed (changes requested):
looking better! next, let's break out the mk2 stuff from the lut files, since those are totally separate.  Again, the goal is to keep the new mk2 code some distance away from the existing code where we can, that will make maintenance easier.

jclsn commented on lib/xwax/lut.c:
```diff
@@ -91,7 +148,26 @@ void lut_push(struct lut *lut, unsigned int timecode)
+    hash = HASH110(timecode);
+    slot->next = lut->table[hash];
+    lut->table[hash] = slot_no;
+}
```
Put into lut_mk2.c

jclsn commented on lib/xwax/timecoder.c:
```diff
@@ -105,6 +109,57 @@ static struct timecode_def timecodes[] = {
         .length = 2110000,
         .safe = 907000,
     },
+    {
+        .name = "traktor_mk2_a",
+        .desc = "Traktor Scratch MK2, side A",
+        .resolution = 2500,
+        .flags = TRAKTOR_MK2,
```
Alright

jclsn committed:
xwax: Implement LUT loading/storing feature

Since the current LUT implementation is designed for Serato-style timecodes
with short LFSRs and not suitable for the Traktor MK2s with its 110-bit LFSR
and up to four million states, the computation of the hash map is taking
quite long (~10 seconds for the CD).

This is a dirty workaround of improving the startup time by not having
to wait for the LUT build every time. I am not happy with this solution,
but it does make the feature usable for now.

In the long run the we should look into pre-computing a Minimum Perfect
Hash Table, which should reduce the size of the LUT drastically.

jclsn commented on lib/xwax/timecoder_mk2.c:
```diff
@@ -0,0 +1,462 @@
+#include <limits.h>
```
Sure

jclsn commented on lib/xwax/lut.c:
```diff
@@ -107,5 +183,27 @@ unsigned int lut_lookup(struct lut *lut, unsigned int timecode)
+
+slot_no_t lut_lookup_mk2(struct lut_mk2 *lut, mk2bits_t *timecode)
+{
+    unsigned int hash;
```
Moved

ywwg reviewed (approved):
this is great! very excited to get this in

jclsn commented:
So I did a last round of cleanups and improved the comments. 

Regarding the MK2-specific structs in the `timecoder.h`, I don't see how to cleanly move those to the `timecoder_mk2.h` without changing the architecture of xwax. Leaving them there now.

Ready to merge from my end!

Swiftb0y commented on lib/xwax/types.h:
```diff
@@ -0,0 +1,173 @@
+
+static inline u128 u128_lshift(u128 a, uint32_t n)
+{
+    u128 result;
+
+    if (n >= 128) {
+        result.high = 0LL;
+        result.low = 0LL;
+    } else if (n >= 64) {
+        result.high = a.low << (n - 64);
+        result.low = 0LL;
+    } else {
+        result.high = (a.high << n) | (a.low >> (64 - n));
+        result.low = a.low << n;
+    }
+
+    return result;
```
small style nit (I'm not a C expert, just transferring my C++ knowledge), but this would feel more robust (as its unlikely to accidentally create an uninitialized object).
 
```suggestion
    if (n >= 128) {
        return U128_ZERO;
    } else if (n >= 64) {
        return U128(a.low << (n - 64), 0LL);
    } else {
        return U128((a.high << n) | (a.low >> (64 - n)), a.low << n);
    }
```

jclsn replied:
Good call! I changed all the functions accordingly, saving a bunch of lines.

Swiftb0y commented on lib/xwax/timecoder.c:
```diff
@@ -603,8 +751,16 @@ void timecoder_submit(struct timecoder *tc, signed short *pcm, size_t npcm)
+            process_sample(tc, primary, secondary);
+
+            /* Display the derivative in the monitor */
+            update_monitor(tc, tc->primary.mk2.deriv_scaled << 1,
+                    tc->secondary.mk2.deriv_scaled << 1);
```
why the left bitshift?

jclsn replied:
That is just a multiplication by two. It makes the signal fit nicely into the scope window.

jclsn replied:
Changed and added an explanation

Swiftb0y reviewed (commented):
couple thoughts. not an expert on C nor xwax though

Swiftb0y reviewed (commented):
couple more nits, rest LGTM.

jclsn commented on lib/xwax/types.h:
```diff
@@ -0,0 +1,130 @@
+
+/*
+ * Bitwise NOT (negation) of a u128 value.
+ */
```
Changed

jclsn commented on lib/xwax/lut_mk2.h:
```diff
@@ -0,0 +1,25 @@
+
+typedef u128 mk2bits_t;
+
+struct slot_mk2 {
+    mk2bits_t timecode;
+    slot_no_t next; /* next slot with the same hash */
+};
+
+struct lut_mk2 {
+    struct slot_mk2 *slot;
+    slot_no_t *table, /* hash -> slot lookup */
+        avail; /* next available slot */
+};
```
It's a hash table that allows collisions in combination with a linked list. In C there is no `std::unordered_list`, so you have to write one yourself. This is Marks implementation. I just needed it to be able to hold an unsigned 128-bit type for the MK2s. Tbh it took me some time to wrap my head around this as well. 

First think of an LFSR having like a lot of unique states it can be in. The LUT is built by shifting the LFSR and assigning each state (called `timecode` here), an index (the slot number). Then the unique state is being hashed with a very basic hash function that doesn't guarantee unique hashes. The resulting hash is stored in the `table` array and holds the slot number as value.

In case the hash function produces the same hash again, it writes the slot number from before in `slot->next` and writes the new slot number in place. That is the linked list. It's a compromise between accuracy and performance. So you do you quick hash and then do linear search for 20 items or so, which (in theory) is a faster lookup than using unique hashes. These days with our modern processors, none of this matters anymore, but Mark wrote this 20 years ago.

```
void lut_push(struct lut *lut, bits_t timecode)               
{                                                             
    unsigned int hash;                                        
    slot_no_t slot_no;                                        
    struct slot *slot;                                        
                                                              
    slot_no = lut->avail++;               // take the next available slot 
                                                          // (and post-increment for the next function call already) 
                                                              
    slot = &lut->slot[slot_no];         // slot points the free slot                          
    slot->timecode = timecode;     // store the LFSR state in the free slot           
                                                              
    hash = HASH(timecode);           // hash the LFSR state (create the key)       
    slot->next = lut->table[hash];   // take the current slot number (value) under the hash (key) 
                                                           // and write it to slot->next, basically creating a linked list 
                                                           
    lut->table[hash] = slot_no;        // store the new slot number as value for the key
}                                                                                                                                                                                                      
```

To lookup, you hash the LFSR state again and traverse down the linked list, until you find a match and extract she slot number

```c 
slot_no_t lut_lookup(struct lut *lut, bits_t timecode)
{
    unsigned int hash;
    slot_no_t slot_no;
    struct slot *slot;

    hash = HASH(timecode);                    // Hash the currently found LFSR state
    slot_no = lut->table[hash];                 // Get the slot number (value) under the hash (key)

    while (slot_no != NO_SLOT) {             // Search until you find an empty slot (NO_SLOT = -1)
        slot = &lut->slot[slot_no];              // Search the slot under the current slot number
        if (slot->timecode == timecode)   // If you find a match, return it
            return slot_no;
        slot_no = slot->next;                       // else traverse down the linked list and try the next 
                                                                   // slot number under the same hash
    }

    return (slot_no_t)-1;                            // if nothing is found, return NO_SLOT
}
```

jclsn commented on lib/xwax/lut_mk2.c:
```diff
@@ -0,0 +1,117 @@
+    for (n = 0; n < hashes; n++)
+        lut->table[n] = NO_SLOT;
+
+    lut->avail = 0;
```
I think it's just a matter of convention. In C this is quite a normal thing to do. That being said, this is C code, so I didn't try to make it adhere to any C++ guidelines.

Swiftb0y replied:
Yeah thats fine, I also recognize how this is the style the rest of the xwax code follows, it was just a source of confusion for me personally.

Swiftb0y reviewed (approved):
LGTM. Thank you.
